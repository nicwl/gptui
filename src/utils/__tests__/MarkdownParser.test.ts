/**
 * Tests for MarkdownParser
 * Refactored to compare entire ASTs via snapshots for clarity.
 */

import { MarkdownParser, MarkdownASTNode } from '../MarkdownParser';
import { MarkdownTokenizer } from '../SimpleMarkdownTokenizer';

function parseMarkdown(text: string): MarkdownASTNode {
  const tokenizer = new MarkdownTokenizer();
  const parser = new MarkdownParser();
  for (const char of text) {
    const tokens = tokenizer.accept(char);
    for (const token of tokens) parser.accept(token);
  }
  const finalTokens = tokenizer.flush();
  for (const token of finalTokens) parser.accept(token);
  parser.flush();
  return parser.getAST();
}

function lines(parts: string[]): string { return parts.join('\n'); }

describe('MarkdownParser (AST snapshots)', () => {
  test.each([
    { name: 'empty document', input: '' },
    { name: 'simple paragraph', input: 'Hello world' },
    { name: 'multiple paragraphs', input: lines(['Paragraph 1', '', 'Paragraph 2']) },
    { name: 'ATX heading level 1', input: '# Heading 1' },
    { name: 'multiple heading levels', input: lines(['# H1', '## H2', '### H3']) },
    { name: 'fenced code block', input: '```\ncode content\n```' },
    { name: 'fenced code block with language', input: '```javascript\nconst x = 1;\n```' },
    { name: 'inline code', input: 'Some `inline code` here' },
    { name: 'bold text', input: '**bold text**' },
    { name: 'italic text', input: '*italic text*' },
    { name: 'nested emphasis', input: '**bold *and italic* text**' },
    { name: 'simple link', input: '[link text](https://example.com)' },
    { name: 'link with emphasis', input: '[**bold link**](url)' },
    { name: 'non-link brackets', input: 'This [is not a link] and neither is [this one' },
    { name: 'incomplete link syntax', input: '[link text without url] and [another](incomplete' },
    { name: 'blockquote simple', input: lines(['> This is a quote', '> that continues.']) },
    { name: 'blockquote with emphasis and link', input: '> *emphasis* and [link](https://example.com)' },
    { name: 'horizontal rule --- between paragraphs', input: lines(['First', '---', 'Second']) },
    { name: 'horizontal rule *** only', input: '***' },
    { name: 'triple-asterisk strong emphasis (with nested em)', input: '***very important*** and ***nested *em* inside***' },
    { name: 'mixed content', input: lines(['# Heading', '', 'This is **bold** and *italic* with `code`.', '', '```js', 'console.log("hello");', '```']) },
    { name: 'flush with open formatting', input: '**incomplete bold' },
    { name: 'only whitespace', input: '   \n  \t  ' },
    { name: 'malformed markdown', input: '**bold without closing *italic without closing [link without closing' },
    // Lists
    { name: 'unordered list (simple)', input: lines(['- First item', '- Second item', '- Third item']) },
    { name: 'ordered list (simple)', input: lines(['1. First item', '2. Second item', '3. Third item']) },
    { name: 'nested unordered list', input: lines(['- First item', '  - Nested item', '  - Another nested item', '- Second item', '  - More nesting', '    - Deep nesting']) },
    { name: 'mixed nested lists', input: lines(['1. First ordered item', '   - Unordered nested', '   - Another unordered', '2. Second ordered item', '   1. Nested ordered', '   2. Another nested ordered']) },
    { name: 'list items with formatting', input: lines(['- **Bold item** with normal text', '- *Italic item* and `code`', '- [Link item](https://example.com) with description', '1. **Bold ordered** item', '2. Item with `inline code` snippet']) },
    { name: 'complex nested lists with formatting', input: lines(['1. **Main task**: Complete the project', '   - Subtask with *emphasis*', '   - Another subtask with `code`', '      1. Deep nested **bold** task', '      2. Another deep task with [link](https://test.com)', '2. **Second main task**: Review and test', '   - Review `codebase`', '   - Test **critical paths**']) },
    { name: 'list item with multiple paragraphs', input: lines(['- First item', '', '  Second paragraph of first item', '', '- Second item with single paragraph']) },
    // Complex real world content
    { name: 'complex team update message', input: lines([
      '# Hello, Team! ðŸ‘‹',
      '',
      "I hope you're all doing well. Here are the updates for **this week**:",
      '',
      '## âœ… Completed Tasks',
      '- Finished the **homepage redesign**',
      '- Deployed the **v2.3 update**',
      '- Fixed the login authentication bug',
      '',
      '## ðŸš€ Upcoming Goals',
      '1. Launch the **marketing campaign**',
      '2. Test the **mobile app beta**',
      '3. Prepare **quarterly report**',
      '',
      '---',
      '',
      '> **Reminder:** The next team meeting is on **Monday at 10 AM**.  ',
      '> Please bring your progress reports.',
      '',
      'Thanks,  ',
      '**Alex**',
    ]) },
    // Code block newlines
    { name: 'code block preserves newlines', input: '```\nfunction test() {\n  console.log("hello");\n  return true;\n}\n```' },
    { name: 'code block preserves empty lines', input: '```\nline1\n\nline3\n\nline5\n```' },
    // New formatting support
    { name: 'underscore italic', input: '_text_' },
    { name: 'underscore strong', input: '__text__' },
    { name: 'triple underscore strong emphasis', input: '___text___' },
    { name: 'mixed bold+italic **_text_**', input: '**_text_**' },
    { name: 'mixed bold+italic _**text**_', input: '_**text**_' },
    { name: 'triple asterisk yields bold+italic', input: '***text***' },
    { name: 'strikethrough', input: '~~text~~' },
    { name: 'mixed strong+strike', input: '**~~gone~~** and ~~**gone too**~~' },
  ])('$name', ({ input }) => {
      const ast = parseMarkdown(input);
    expect(ast).toMatchSnapshot();
  });
});
