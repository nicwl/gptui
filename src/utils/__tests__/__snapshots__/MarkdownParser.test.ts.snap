// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownParser (AST snapshots) ATX heading level 1 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "Heading 1",
          "type": "text",
        },
      ],
      "metadata": {
        "level": 1,
      },
      "type": "heading",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) blockquote simple 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": " This is a quote  that continues.",
          "type": "text",
        },
      ],
      "type": "blockquote",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) blockquote with emphasis and link 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": " ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "emphasis",
              "type": "text",
            },
          ],
          "type": "emphasis",
        },
        {
          "content": " and ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "link",
              "type": "text",
            },
          ],
          "metadata": {
            "url": "https://example.com",
          },
          "type": "link",
        },
      ],
      "type": "blockquote",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) bold text 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "bold text",
              "type": "text",
            },
          ],
          "type": "strong",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) code block preserves empty lines 1`] = `
{
  "children": [
    {
      "content": "line1

line3

line5",
      "type": "code_block",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) code block preserves newlines 1`] = `
{
  "children": [
    {
      "content": "function test() {
  console.log("hello");
  return true;
}",
      "type": "code_block",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) complex nested lists with formatting 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "Main task",
              "type": "text",
            },
          ],
          "type": "strong",
        },
        {
          "content": ": Complete the project",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": 1,
        "ordered": true,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Subtask with ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "emphasis",
              "type": "text",
            },
          ],
          "type": "emphasis",
        },
      ],
      "metadata": {
        "depth": 2,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Another subtask with ",
          "type": "text",
        },
        {
          "content": "code",
          "type": "code_inline",
        },
      ],
      "metadata": {
        "depth": 2,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Deep nested ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "bold",
              "type": "text",
            },
          ],
          "type": "strong",
        },
        {
          "content": " task",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 4,
        "number": 1,
        "ordered": true,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Another deep task with ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "link",
              "type": "text",
            },
          ],
          "metadata": {
            "url": "https://test.com",
          },
          "type": "link",
        },
      ],
      "metadata": {
        "depth": 4,
        "number": 2,
        "ordered": true,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "children": [
            {
              "content": "Second main task",
              "type": "text",
            },
          ],
          "type": "strong",
        },
        {
          "content": ": Review and test",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": 2,
        "ordered": true,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Review ",
          "type": "text",
        },
        {
          "content": "codebase",
          "type": "code_inline",
        },
      ],
      "metadata": {
        "depth": 2,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Test ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "critical paths",
              "type": "text",
            },
          ],
          "type": "strong",
        },
      ],
      "metadata": {
        "depth": 2,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) complex team update message 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "Hello, Team! ðŸ‘‹",
          "type": "text",
        },
      ],
      "metadata": {
        "level": 1,
      },
      "type": "heading",
    },
    {
      "children": [
        {
          "content": "I hope you're all doing well. Here are the updates for ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "this week",
              "type": "text",
            },
          ],
          "type": "strong",
        },
        {
          "content": ": ",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "children": [
        {
          "content": "âœ… Completed Tasks",
          "type": "text",
        },
      ],
      "metadata": {
        "level": 2,
      },
      "type": "heading",
    },
    {
      "children": [
        {
          "content": "Finished the ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "homepage redesign",
              "type": "text",
            },
          ],
          "type": "strong",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Deployed the ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "v2.3 update",
              "type": "text",
            },
          ],
          "type": "strong",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Fixed the login authentication bug",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "ðŸš€ Upcoming Goals",
          "type": "text",
        },
      ],
      "metadata": {
        "level": 2,
      },
      "type": "heading",
    },
    {
      "children": [
        {
          "content": "Launch the ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "marketing campaign",
              "type": "text",
            },
          ],
          "type": "strong",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": 1,
        "ordered": true,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Test the ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "mobile app beta",
              "type": "text",
            },
          ],
          "type": "strong",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": 2,
        "ordered": true,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Prepare ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "quarterly report",
              "type": "text",
            },
          ],
          "type": "strong",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": 3,
        "ordered": true,
      },
      "type": "list_item",
    },
    {
      "type": "hr",
    },
    {
      "children": [
        {
          "content": " ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "Reminder:",
              "type": "text",
            },
          ],
          "type": "strong",
        },
        {
          "content": " The next team meeting is on ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "Monday at 10 AM",
              "type": "text",
            },
          ],
          "type": "strong",
        },
        {
          "content": ".    Please bring your progress reports. ",
          "type": "text",
        },
      ],
      "type": "blockquote",
    },
    {
      "children": [
        {
          "content": "Thanks,  ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "Alex",
              "type": "text",
            },
          ],
          "type": "strong",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) empty document 1`] = `
{
  "children": [],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) fenced code block 1`] = `
{
  "children": [
    {
      "content": "code content",
      "type": "code_block",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) fenced code block with language 1`] = `
{
  "children": [
    {
      "content": "const x = 1;",
      "metadata": {
        "language": "javascript",
      },
      "type": "code_block",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) flush with open formatting 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "**incomplete bold",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) horizontal rule *** only 1`] = `
{
  "children": [
    {
      "type": "hr",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) horizontal rule --- between paragraphs 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "First",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "type": "hr",
    },
    {
      "children": [
        {
          "content": "Second",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) incomplete link syntax 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "[link text without url] and [another](incomplete",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) inline code 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "Some ",
          "type": "text",
        },
        {
          "content": "inline code",
          "type": "code_inline",
        },
        {
          "content": " here",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) italic text 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "italic text",
              "type": "text",
            },
          ],
          "type": "emphasis",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) link with emphasis 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "content": "bold link",
                  "type": "text",
                },
              ],
              "type": "strong",
            },
          ],
          "metadata": {
            "url": "url",
          },
          "type": "link",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) list item with multiple paragraphs 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "First item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "  Second paragraph of first item ",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "children": [
        {
          "content": "Second item with single paragraph",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) list items with formatting 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "Bold item",
              "type": "text",
            },
          ],
          "type": "strong",
        },
        {
          "content": " with normal text",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "children": [
            {
              "content": "Italic item",
              "type": "text",
            },
          ],
          "type": "emphasis",
        },
        {
          "content": " and ",
          "type": "text",
        },
        {
          "content": "code",
          "type": "code_inline",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "children": [
            {
              "content": "Link item",
              "type": "text",
            },
          ],
          "metadata": {
            "url": "https://example.com",
          },
          "type": "link",
        },
        {
          "content": " with description",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "children": [
            {
              "content": "Bold ordered",
              "type": "text",
            },
          ],
          "type": "strong",
        },
        {
          "content": " item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": 1,
        "ordered": true,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Item with ",
          "type": "text",
        },
        {
          "content": "inline code",
          "type": "code_inline",
        },
        {
          "content": " snippet",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": 2,
        "ordered": true,
      },
      "type": "list_item",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) malformed markdown 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "**bold without closing *italic without closing [link without closing",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) mixed bold+italic **_text_** 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "content": "text",
                  "type": "text",
                },
              ],
              "type": "emphasis",
            },
          ],
          "type": "strong",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) mixed bold+italic _**text**_ 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "content": "text",
                  "type": "text",
                },
              ],
              "type": "strong",
            },
          ],
          "type": "emphasis",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) mixed content 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "Heading",
          "type": "text",
        },
      ],
      "metadata": {
        "level": 1,
      },
      "type": "heading",
    },
    {
      "children": [
        {
          "content": "This is ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "bold",
              "type": "text",
            },
          ],
          "type": "strong",
        },
        {
          "content": " and ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "italic",
              "type": "text",
            },
          ],
          "type": "emphasis",
        },
        {
          "content": " with ",
          "type": "text",
        },
        {
          "content": "code",
          "type": "code_inline",
        },
        {
          "content": ". ",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": "console.log("hello");",
      "metadata": {
        "language": "js",
      },
      "type": "code_block",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) mixed nested lists 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "First ordered item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": 1,
        "ordered": true,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Unordered nested",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 2,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Another unordered",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 2,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Second ordered item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": 2,
        "ordered": true,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Nested ordered",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 2,
        "number": 1,
        "ordered": true,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Another nested ordered",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 2,
        "number": 2,
        "ordered": true,
      },
      "type": "list_item",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) mixed strong+strike 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "content": "gone",
                  "type": "text",
                },
              ],
              "type": "strikethrough",
            },
          ],
          "type": "strong",
        },
        {
          "content": " and ",
          "type": "text",
        },
        {
          "children": [
            {
              "children": [
                {
                  "content": "gone too",
                  "type": "text",
                },
              ],
              "type": "strong",
            },
          ],
          "type": "strikethrough",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) multiple heading levels 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "H1",
          "type": "text",
        },
      ],
      "metadata": {
        "level": 1,
      },
      "type": "heading",
    },
    {
      "children": [
        {
          "content": "H2",
          "type": "text",
        },
      ],
      "metadata": {
        "level": 2,
      },
      "type": "heading",
    },
    {
      "children": [
        {
          "content": "H3",
          "type": "text",
        },
      ],
      "metadata": {
        "level": 3,
      },
      "type": "heading",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) multiple paragraphs 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "Paragraph 1 ",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "children": [
        {
          "content": "Paragraph 2",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) nested emphasis 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "bold ",
              "type": "text",
            },
            {
              "children": [
                {
                  "content": "and italic",
                  "type": "text",
                },
              ],
              "type": "emphasis",
            },
            {
              "content": " text",
              "type": "text",
            },
          ],
          "type": "strong",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) nested unordered list 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "First item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Nested item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 2,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Another nested item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 2,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Second item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "More nesting",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 2,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Deep nesting",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 3,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) non-link brackets 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "This [is not a link] and neither is [this one",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) only whitespace 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "     	  ",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) ordered list (simple) 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "First item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": 1,
        "ordered": true,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Second item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": 2,
        "ordered": true,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Third item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": 3,
        "ordered": true,
      },
      "type": "list_item",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) simple link 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "link text",
              "type": "text",
            },
          ],
          "metadata": {
            "url": "https://example.com",
          },
          "type": "link",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) simple paragraph 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "Hello world",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) strikethrough 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "text",
              "type": "text",
            },
          ],
          "type": "strikethrough",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) triple asterisk yields bold+italic 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "text",
              "type": "text",
            },
          ],
          "type": "strong_emphasis",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) triple underscore strong emphasis 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "text",
              "type": "text",
            },
          ],
          "type": "strong_emphasis",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) triple-asterisk strong emphasis (with nested em) 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "very important",
              "type": "text",
            },
          ],
          "type": "strong_emphasis",
        },
        {
          "content": " and ***nested ",
          "type": "text",
        },
        {
          "children": [
            {
              "content": "em",
              "type": "text",
            },
          ],
          "type": "emphasis",
        },
        {
          "content": " inside***",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) underscore italic 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "text",
              "type": "text",
            },
          ],
          "type": "emphasis",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) underscore strong 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "text",
              "type": "text",
            },
          ],
          "type": "strong",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "document",
}
`;

exports[`MarkdownParser (AST snapshots) unordered list (simple) 1`] = `
{
  "children": [
    {
      "children": [
        {
          "content": "First item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Second item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
    {
      "children": [
        {
          "content": "Third item",
          "type": "text",
        },
      ],
      "metadata": {
        "depth": 1,
        "number": undefined,
        "ordered": false,
      },
      "type": "list_item",
    },
  ],
  "type": "document",
}
`;
